### PR TITLE
Adding first draft for review

### PR DESCRIPTION
**Pull request by dkaranjai**

### CODE REVIEW

**Comment by dkaranjai** [comment]
# k8_dpdk_add_remove_services

[root@wp06 ptf-tests]# ptf --test-dir tests/ k8_dpdk_add_remove_services --pypath $PWD --test-params="kube_iperf_path='/root/go/src/p4-k8s/kubernetes-iperf3'" --platform=dummy
/usr/local/bin/ptf:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using packet manipulation module: ptf.packet_scapy
k8_dpdk_add_remove_services.K8_DPDK_services ... 2/15/2023,23:43:33  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/15/2023,23:43:33  PASSED: Successfully executed setup.sh
2/15/2023,23:43:33  INFO: Checking if service:iperf3-server is running or not
2/15/2023,23:43:33  INFO: Service iperf3-server running
2/15/2023,23:43:33  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/15/2023,23:43:33  PASSED: Successfully executed cleanup.sh
2/15/2023,23:43:33  INFO: Checking if service:iperf3-server is running or not
2/15/2023,23:43:33  INFO: Service iperf3-server not running
2/15/2023,23:43:33  INFO: Test has PASSED
ok

----------------------------------------------------------------------
Ran 1 test in 10.931s

OK

File: tests/k8_dpdk_add_remove_services.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        # Execute ./steps/setup.sh from https://github.com/Pharb/kubernetes-iperf3.git
+        if not k8_utils.execute_iperf_setup(self.kube_iperf_path):
+            self.result.addFailure(self, sys.exc_info())
+            self.fail("Failed to run setup.sh")
+
+        # Check if iperf3-server is running
+        if not k8_utils.check_service_status("iperf3-server"):
+            self.result.addFailure(self, sys.exc_info())
+            self.fail("iperf3-server not running")
+
```
Comment (djayakum) [inline tests/k8_dpdk_add_remove_services.py:65]: Traffic test is missing. iperf client is not started

File: tests/k8_dpdk_add_remove_services.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        BaseTest.setUp(self)
+        self.result = unittest.TestResult()
+        config[
+            "relax"
+        ] = True  # for verify_packets to ignore other packets received at the interface
+        test_params = test_params_get()
+        self.kube_iperf_path = test_params["kube_iperf_path"]
+
+    def runTest(self):
+        # Execute ./steps/setup.sh from https://github.com/Pharb/kubernetes-iperf3.git
```
Comment (djayakum) [inline tests/k8_dpdk_add_remove_services.py:56]: Do you assume that this repo is cloned before the start of this test?

**Review by djayakum — CHANGES_REQUESTED** [review]

File: tests/k8_dpdk_add_remove_services.py (hunk 2 above)
Comment (dkaranjai) [reply tests/k8_dpdk_add_remove_services.py:56]: Will check, if not, will add another util to do this.

File: common/utils/k8_utils.py [hunk 3]
```diff
@@ -165,3 +166,72 @@ def ping_and_verify_no_packet_loss(pod_name, dest_ip, count=10):
...
+
+    log.info(ret)
+
+    if "Server is running" in ret:
+        log.passed("Successfully executed setup.sh")
+        return True
+
+    log.failed("Failed to execute setup.sh")
+    return False
+
```
Comment (stadju) [inline common/utils/k8_utils.py:213]: Is there a util for run.sh?

Comment (dkaranjai) [reply common/utils/k8_utils.py:213]: not yet, these libs are created for the test case add/remove services. will create these utils as and when required.

**Comment by dkaranjai** [comment]
## k8_dpdk_add_remove_services

[root@wp06 ptf-tests]# ptf --test-dir tests/ k8_dpdk_add_remove_services --pypath $PWD --test-params="kube_iperf_path='/tmp/kubernetes-iperf3'" --platform=dummy                                        /usr/local/bin/ptf:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using packet manipulation module: ptf.packet_scapy
k8_dpdk_add_remove_services.K8_DPDK_services ... 
### 2/16/2023,22:7:06  PASSED: Git clone github.com/Pharb/kubernetes-iperf3.git Successful
2/16/2023,22:7:17  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/16/2023,22:7:17  PASSED: Successfully executed setup.sh
2/16/2023,22:7:17  INFO: Checking if service:iperf3-server is running or not
2/16/2023,22:7:17  INFO: Service iperf3-server running
2/16/2023,22:7:17  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/16/2023,22:7:17  PASSED: Successfully executed cleanup.sh
2/16/2023,22:7:17  INFO: Checking if service:iperf3-server is running or not
2/16/2023,22:7:17  INFO: Service iperf3-server not running
2/16/2023,22:7:17  INFO: Test has PASSED
ok

----------------------------------------------------------------------
Ran 1 test in 11.599s

OK

**Comment by dkaranjai** [comment]
# k8_dpdk_service_traffic_test

[root@wp06 ptf-tests]# ptf --test-dir tests/ k8_dpdk_service_traffic_test --pypath $PWD --test-params="kube_iperf_path='/tmp/kubernetes-iperf3'" --platform=dummy
/usr/local/bin/ptf:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using packet manipulation module: ptf.packet_scapy
k8_dpdk_service_traffic_test.K8_DPDK_service_traffic ... 2/22/2023,22:55:50  PASSED: Git clone github.com/Pharb/kubernetes-iperf3.git Successful
2/22/2023,22:56:00  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/22/2023,22:56:00  PASSED: Successfully executed setup.sh
2/22/2023,22:56:00  INFO: Checking if service:iperf3-server is running or not
2/22/2023,22:56:00  INFO: Service iperf3-server running
2/22/2023,22:56:05  INFO: Starting iperf client
Unable to use a TTY - input is not a terminal or the right kind of file
2/22/2023,22:56:16  INFO: Client on 10.233.132.110:  Connecting to host iperf3-server, port 5201
Client on 10.233.132.110:  [  5] local 10.244.0.20 port 56358 connected to 10.96.137.230 port 5201
Client on 10.233.132.110:  [ ID] Interval           Transfer     Bitrate         Retr  Cwnd
Client on 10.233.132.110:  [  5]   0.00-1.00   sec  70.8 MBytes   594 Mbits/sec   48   1.10 MBytes
Client on 10.233.132.110:  [  5]   1.00-2.00   sec  68.8 MBytes   577 Mbits/sec    0   1.21 MBytes
Client on 10.233.132.110:  [  5]   2.00-3.00   sec  68.8 MBytes   577 Mbits/sec    0   1.30 MBytes
Client on 10.233.132.110:  [  5]   3.00-4.00   sec  67.5 MBytes   566 Mbits/sec    0   1.37 MBytes
Client on 10.233.132.110:  [  5]   4.00-5.00   sec  67.5 MBytes   566 Mbits/sec    3   1.02 MBytes
Client on 10.233.132.110:  [  5]   5.00-6.00   sec  68.8 MBytes   577 Mbits/sec    0   1.08 MBytes
Client on 10.233.132.110:  [  5]   6.00-7.00   sec  68.8 MBytes   577 Mbits/sec    0   1.12 MBytes
Client on 10.233.132.110:  [  5]   7.00-8.00   sec  66.2 MBytes   556 Mbits/sec    0   1.15 MBytes
Client on 10.233.132.110:  [  5]   8.00-9.00   sec  68.8 MBytes   577 Mbits/sec    0   1.18 MBytes
Client on 10.233.132.110:  [  5]   9.00-10.00  sec  67.5 MBytes   566 Mbits/sec    0   1.22 MBytes
Client on 10.233.132.110:  - - - - - - - - - - - - - - - - - - - - - - - - -
Client on 10.233.132.110:  [ ID] Interval           Transfer     Bitrate         Retr
Client on 10.233.132.110:  [  5]   0.00-10.00  sec   683 MBytes   573 Mbits/sec   51             sender
Client on 10.233.132.110:  [  5]   0.00-10.02  sec   682 MBytes   571 Mbits/sec                  receiver
Client on 10.233.132.110:
Client on 10.233.132.110:  iperf Done.


2/22/2023,22:56:16  INFO: Sent 683 MBytes, Received 682 MBytes
2/22/2023,22:56:16  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/22/2023,22:56:16  PASSED: Successfully executed cleanup.sh
2/22/2023,22:56:16  INFO: Checking if service:iperf3-server is running or not
2/22/2023,22:56:16  INFO: Service iperf3-server not running
2/22/2023,22:56:16  INFO: Test has PASSED
ok

----------------------------------------------------------------------
Ran 1 test in 27.225s

OK

**Comment by dkaranjai** [comment]
#  k8_dpdk_add_remove_services_n_times

[root@wp06 ptf-tests]# ptf --test-dir tests/ k8_dpdk_add_remove_services_n_times --pypath $PWD --test-params="kube_iperf_path='/tmp/kubernetes-iperf3';no_of_times='4'" --platform=dummy
/usr/local/bin/ptf:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using packet manipulation module: ptf.packet_scapy
k8_dpdk_add_remove_services_n_times.K8_DPDK_services ... 2/22/2023,23:10:15  PASSED: Git clone github.com/Pharb/kubernetes-iperf3.git Successful
2/22/2023,23:10:15  INFO: Add/Remove services: Iteration 1
2/22/2023,23:10:26  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/22/2023,23:10:26  PASSED: Successfully executed setup.sh
2/22/2023,23:10:26  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:26  INFO: Service iperf3-server running
2/22/2023,23:10:26  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/22/2023,23:10:26  PASSED: Successfully executed cleanup.sh
2/22/2023,23:10:26  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:26  INFO: Service iperf3-server not running
2/22/2023,23:10:26  INFO: Add/Remove services: Iteration 2
2/22/2023,23:10:37  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/22/2023,23:10:37  PASSED: Successfully executed setup.sh
2/22/2023,23:10:37  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:37  INFO: Service iperf3-server running
2/22/2023,23:10:37  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/22/2023,23:10:37  PASSED: Successfully executed cleanup.sh
2/22/2023,23:10:37  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:37  INFO: Service iperf3-server not running
2/22/2023,23:10:37  INFO: Add/Remove services: Iteration 3
2/22/2023,23:10:47  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/22/2023,23:10:47  PASSED: Successfully executed setup.sh
2/22/2023,23:10:47  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:48  INFO: Service iperf3-server running
2/22/2023,23:10:48  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/22/2023,23:10:48  PASSED: Successfully executed cleanup.sh
2/22/2023,23:10:48  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:48  INFO: Service iperf3-server not running
2/22/2023,23:10:48  INFO: Add/Remove services: Iteration 4
2/22/2023,23:10:58  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/22/2023,23:10:58  PASSED: Successfully executed setup.sh
2/22/2023,23:10:58  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:58  INFO: Service iperf3-server running
2/22/2023,23:10:59  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/22/2023,23:10:59  PASSED: Successfully executed cleanup.sh
2/22/2023,23:10:59  INFO: Checking if service:iperf3-server is running or not
2/22/2023,23:10:59  INFO: Service iperf3-server not running
2/22/2023,23:10:59  INFO: Test has PASSED
ok

----------------------------------------------------------------------
Ran 1 test in 44.341s

OK

**Review by stadju — COMMENTED** [review]
I updated the test procedure for add/remove service:
1. Follow wiki instruction to setup K8S and create services and make sure services are running

https://github.com/ipdk-io/k8s-infra-offload/blob/main/README.md

steps/setup.sh

kubectl get svc -A should show iperf3-server

2. run steps/cleanup.sh

kubectl get svc -A should not show iperf3-server

3. Create services again

steps/setup.sh

4. Run traffic test: steps/run.sh

5. Verify non-zero bytes transferred

6. Re-run step 2

**Comment by dkaranjai** [comment]
> ptf --test-dir tests/ k8_dpdk_add_remove_services --pypath $PWD --test-params="kube_iperf_path='/tmp/kubernetes-iperf3'" --platform=dummy


@stadju -- fixed this. updated log:

[root@wp06 ptf-tests]# ptf --test-dir tests/ k8_dpdk_add_remove_services --pypath $PWD --test-params="kube_iperf_path='/tmp/kubernetes-iperf3'" --platform=dummy
/usr/local/bin/ptf:19: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
Using packet manipulation module: ptf.packet_scapy
k8_dpdk_add_remove_services.K8_DPDK_services ... 2/23/2023,21:38:00  PASSED: Git clone github.com/Pharb/kubernetes-iperf3.git Successful
2/23/2023,21:38:11  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/23/2023,21:38:11  PASSED: Successfully executed setup.sh
2/23/2023,21:38:11  INFO: Checking if service:iperf3-server is running or not
2/23/2023,21:38:11  INFO: Service iperf3-server running
2/23/2023,21:38:11  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/23/2023,21:38:11  PASSED: Successfully executed cleanup.sh
2/23/2023,21:38:11  INFO: Checking if service:iperf3-server is running or not
2/23/2023,21:38:11  INFO: Service iperf3-server not running
2/23/2023,21:38:22  INFO: deployment.apps/iperf3-server-deployment created
service/iperf3-server created
daemonset.apps/iperf3-clients created
Waiting for iperf3 server to start...
Waiting for iperf3 server to start...
Server is running


2/23/2023,21:38:22  PASSED: Successfully executed setup.sh
2/23/2023,21:38:22  INFO: Checking if service:iperf3-server is running or not
2/23/2023,21:38:22  INFO: Service iperf3-server running
2/23/2023,21:38:27  INFO: Starting iperf client
Unable to use a TTY - input is not a terminal or the right kind of file
2/23/2023,21:38:38  INFO: Client on 10.233.132.110:  Connecting to host iperf3-server, port 5201
Client on 10.233.132.110:  [  5] local 10.244.0.39 port 60124 connected to 10.96.34.47 port 5201
Client on 10.233.132.110:  [ ID] Interval           Transfer     Bitrate         Retr  Cwnd
Client on 10.233.132.110:  [  5]   0.00-1.00   sec  71.4 MBytes   599 Mbits/sec   46   1.09 MBytes
Client on 10.233.132.110:  [  5]   1.00-2.00   sec  67.5 MBytes   566 Mbits/sec    0   1.20 MBytes
Client on 10.233.132.110:  [  5]   2.00-3.00   sec  67.5 MBytes   566 Mbits/sec    0   1.29 MBytes
Client on 10.233.132.110:  [  5]   3.00-4.00   sec  66.2 MBytes   556 Mbits/sec    0   1.35 MBytes
Client on 10.233.132.110:  [  5]   4.00-5.00   sec  68.8 MBytes   577 Mbits/sec    1   1.00 MBytes
Client on 10.233.132.110:  [  5]   5.00-6.00   sec  67.5 MBytes   566 Mbits/sec    0   1.07 MBytes
Client on 10.233.132.110:  [  5]   6.00-7.00   sec  67.5 MBytes   566 Mbits/sec    0   1.11 MBytes
Client on 10.233.132.110:  [  5]   7.00-8.00   sec  67.5 MBytes   566 Mbits/sec    0   1.14 MBytes
Client on 10.233.132.110:  [  5]   8.00-9.00   sec  68.8 MBytes   577 Mbits/sec    0   1.17 MBytes
Client on 10.233.132.110:  [  5]   9.00-10.00  sec  67.5 MBytes   566 Mbits/sec    0   1.21 MBytes
Client on 10.233.132.110:  - - - - - - - - - - - - - - - - - - - - - - - - -
Client on 10.233.132.110:  [ ID] Interval           Transfer     Bitrate         Retr
Client on 10.233.132.110:  [  5]   0.00-10.00  sec   680 MBytes   571 Mbits/sec   47             sender
Client on 10.233.132.110:  [  5]   0.00-10.02  sec   678 MBytes   568 Mbits/sec                  receiver
Client on 10.233.132.110:
Client on 10.233.132.110:  iperf Done.


2/23/2023,21:38:38  INFO: Sent 680 MBytes, Received 678 MBytes
2/23/2023,21:38:38  INFO: deployment.apps "iperf3-server-deployment" deleted
service "iperf3-server" deleted
daemonset.apps "iperf3-clients" deleted

2/23/2023,21:38:38  PASSED: Successfully executed cleanup.sh
2/23/2023,21:38:38  INFO: Checking if service:iperf3-server is running or not
2/23/2023,21:38:38  INFO: Service iperf3-server not running
2/23/2023,21:38:38  INFO: Test has PASSED
ok

----------------------------------------------------------------------
Ran 1 test in 38.220s

OK

**Review by stadju — APPROVED** [review]

**Pull request merged**